nntrainer/graph/network_graph.cpp

```diff
@@ -426,14 +436,51 @@ void NetworkGraph::backwarding(
   if (clip_weights.empty())
     return;
 
+  // update loss scale
+  auto update_loss_scale = [&](float scale) {
+    ml_logd("set loss scale = %f", scale);
+    for (auto iter = cbegin(); iter != cend(); iter++) {
+      auto &ln = *iter;
+      ln->setLossSacle(scale);
+    }
+    loss_scale = scale;
+  };
+
+  // check first layer's derivative is valid
+  // loss scale is adjusted between 1.0f ~ 256.0f
+  // @todo provide max scale property
+  auto &ln = *(cbegin() + 1);
+  if (loss_scale != 0.0f && !ln->getRunContext().validateDerivatives()) {
+    // It will not apply train results if data is invalid
+    float scale = loss_scale - 0.5f;
+    ml_logd(
+      "Derivative validation failed. Skip applying gradient. loss_scale(%f)",
+      scale);
+    update_loss_scale(scale);
+    return;
+  }
+
   /** calculate the global norm */
   Tensor global_norm_t(
     TensorDim({1u, 1u, 1u, (unsigned int)clip_weights.size()}));
   float *global_norm_data = global_norm_t.getData();
   for (unsigned int idx = 0; idx < clip_weights.size(); idx++) {
     auto const &w = clip_weights[idx];
+
+    if (loss_scale != 0.0f) {
+      w->applyScaler(loss_scale);
+      if (w->getGradient().checkDataValidation(false) == false) {
+        float scale = loss_scale - 0.5f;
```

Comment:
I think 0.5f @here and 2.0f @L496 can be hyper-param.

Reply:
fixed